mava/utils/sebulba_utils.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import queue
+import threading
+import time
+from typing import Any, Dict, List, Sequence, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+from jumanji.types import TimeStep
+from omegaconf import DictConfig
+
+from mava.systems.ppo.types import Params, PPOTransition  # todo: remove the ppo dependencies
+
+
+# Copied from https://github.com/instadeepai/sebulba/blob/main/sebulba/core.py
+class ThreadLifetime:
+    """Simple class for a mutable boolean that can be used to signal a thread to stop."""
+
+    def __init__(self) -> None:
+        self._stop = False
+
+    def should_stop(self) -> bool:
+        return self._stop
+
+    def stop(self) -> None:
+        self._stop = True
+
+
+class Pipeline(threading.Thread):
+    """
+    The `Pipeline` shards trajectories into `learner_devices`,
+    ensuring trajectories are consumed in the right order to avoid being off-policy
+    and limit the max number of samples in device memory at one time to avoid OOM issues.
+    """
+
+    def __init__(self, max_size: int, learner_devices: List[jax.Device], lifetime: ThreadLifetime):
+        """
+        Initializes the pipeline with a maximum size and the devices to shard trajectories across.
+
+        Args:
+            max_size: The maximum number of trajectories to keep in the pipeline.
+            learner_devices: The devices to shard trajectories across.
+        """
+        super().__init__(name="Pipeline")
+        self.learner_devices = learner_devices
+        self.tickets_queue: queue.Queue = queue.Queue()
+        self._queue: queue.Queue = queue.Queue(maxsize=max_size)
+        self.lifetime = lifetime
+
+    def run(self) -> None:
+        """This function ensures that trajectories on the queue are consumed in the right order. The
+        start_condition and end_condition are used to ensure that only 1 thread is processing an
+        item from the queue at one time, ensuring predictable memory usage.
+        """
+        while not self.lifetime.should_stop():
+            try:
+                start_condition, end_condition = self.tickets_queue.get(timeout=1)
```

Comment:
I don't think we want a timeout here

```suggestion
                start_condition, end_condition = self.tickets_queue.get()
```

Reply:
I believe this timeout is necessary because, after the actors shut down, no new tickets are added to the queue, which causes the pipeline to stall.